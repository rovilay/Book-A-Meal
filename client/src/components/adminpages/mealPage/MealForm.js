/* eslint jsx-a11y/label-has-for: 0 */
/* eslint no-alert: 0 */
/* eslint no-restricted-globals: 0 */
import React from 'react';
import PropTypes from 'prop-types';
import FontAwesome from 'react-fontawesome';

const Mealform = props => (
  <div className="card">
    <div id="form-card" className="card-body">
      <div className="form-title" id="form-title">
        {
          (props.isEdit)
          ?
          'Edit Meal'
          :
          'Add Meal'
        }
        <hr />
      </div>

      <form
        id="meal-form"
        className="meal-form"
        onSubmit={(e) => {
          e.preventDefault();
          if (props.isEdit) {
           return props.updateMeal();
          }
          return props.addMeal();
        }}
      >

        <p className="price">
          <label htmlFor="name">Name
          </label>
          <input
            type="text"
            placeholder="Enter meal name"
            name="meal-name"
            id="meal-name"
            required
          />
        </p>

        <p className="price">
          <label htmlFor="price">Unit Price (&#8358;)
          </label>
          <input
            type="number"
            placeholder="Enter price"
            name="price"
            id="price"
            min="0"
            step="50"
            required
          />
        </p>

        <p className="dsc">
          <label htmlFor="dsc">Description
          </label>
          <input
            type="text"
            placeholder="Enter meal description"
            name="dsc"
            id="dsc"
            required
          />
        </p>

        <p>
          <label htmlFor="image">Image
          </label>
          <input
            type="file"
            placeholder="Enter img link"
            name="image"
            id="image"
            onChange={(e) => {
              e.preventDefault();
              if (props.checkFileSize()) {
                props.showUploadBar();
              }
            }}
          />
        </p>
        {
          (props.imageToUpload)
          &&
          <div className="progress">
            <div id="progressBar" className="progressBar">0%</div>
          </div>
        }
        {
          (!props.isEdit)
          &&
          (
            <p className="full">
              <button
                type="submit"
                name="addbtn"
                id="add-btn"
                className="add-btn btn-3"
                disabled={props.disableBtn}
              >
                <FontAwesome
                  name="plus"
                />
                ADD MEAL
              </button>
            </p>
          )
        }
        {
          (props.isEdit)
          &&
          (
            <p>
              <button
                type="submit"
                name="updatebtn"
                id="update-btn"
                className="update-btn btn-3"
              >
                <FontAwesome
                  name="arrow-circle-up"
                />
                  UPDATE MEAL
              </button>

              <button
                type="submit"
                name="deletebtn"
                id="delete-btn"
                className="delete-btn btn-3"
                onClick={(e) => {
                  e.preventDefault();
                  const confirmed = confirm('Are you sure you want to delete this meal?');
                  if (confirmed) {
                    props.deleteMeal(props.mealOnEditId);
                    setTimeout(() => {
                      if (!props.serverRes.success && !props.serverRes.message) {
                        props.notify('Meal was DELETED successfully');
                        props.closeEdit();
                        location.reload();
                      }
                      if (props.serverRes.success === false) {
                        props.notify(props.serverRes.message);
                      }
                    }, 200);
                  }
                }}
              >
                <FontAwesome
                  name="times"
                />
                DELETE MEAL
              </button>

              <button
                type="submit"
                name="backbtn"
                id="back-btn"
                className="back-btn btn-3"
                onClick={(e) => {
                  e.preventDefault();
                  props.closeEdit();
                }}
              >
                <FontAwesome
                  name="arrow-circle-left"
                />
                BACK
              </button>
            </p>
          )
        }
      </form>
    </div>
  </div>
);

Mealform.propTypes = {
  closeEdit: PropTypes.func.isRequired,
  addMeal: PropTypes.func.isRequired,
  updateMeal: PropTypes.func.isRequired,
  deleteMeal: PropTypes.func.isRequired,
  isEdit: PropTypes.bool.isRequired,
  serverRes: PropTypes.object.isRequired,
  mealOnEditId: PropTypes.string.isRequired,
  notify: PropTypes.func.isRequired,
  checkFileSize: PropTypes.func.isRequired,
  imageToUpload: PropTypes.string.isRequired,
  showUploadBar: PropTypes.func.isRequired,
  disableBtn: PropTypes.bool.isRequired
};

export default Mealform;
