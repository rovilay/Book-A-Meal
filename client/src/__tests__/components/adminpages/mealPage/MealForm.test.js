import React from 'react';
import { shallow } from 'enzyme';
import  MealForm from '../../../../components/adminpages/mealPage/MealForm';
import { meals } from '../../../__mockData__/mealMock';


const props = {
  closeEdit: jest.fn(),
  addMeal: jest.fn(),
  handleChange: jest.fn(),
  updateMeal: jest.fn(),
  isEdit: true,
  imageToUpload: 'image.com',
  uploadedImageLink: 'http://image.com',
  uploadImage: jest.fn(),
  disableBtn: false,
  mealOnEdit: {
    ...meals[0],
    description: 'So sweet'
  }
};

describe('MealForm component test', () => {
  const setup = () => shallow(<MealForm {...props} />);

  it('should render MealForm correctly', (done) => {
    const wrapper = setup();

    const event = {
      preventDefault: jest.fn()
    }

    wrapper.find(`.back-btn`).simulate('click', event);
    wrapper.find(`.meal-form`).simulate('submit', event);
    expect(wrapper).toMatchSnapshot();
    expect(props.closeEdit).toHaveBeenCalled();
    expect(props.updateMeal).toHaveBeenCalled();

    done();
  });


  it('should call `addMeal` on meal form submit if not in edit state', (done) => {
    props.isEdit = false;
    const wrapper = setup();

    const event = {
      preventDefault: jest.fn()
    }

    wrapper.find(`.meal-form`).simulate('submit', event);
    expect(wrapper).toMatchSnapshot();
    expect(props.addMeal).toHaveBeenCalled();

    done();
  });


  it('should call `handleChange` on meal form input change', (done) => {
    props.isEdit = false;
    const wrapper = setup();

    const event = {
      preventDefault: jest.fn()
    }

    wrapper.find(`#meal-name`).simulate('change', event);
    wrapper.find(`#dsc`).simulate('change', event);
    wrapper.find(`#price`).simulate('change', event);
    expect(wrapper).toMatchSnapshot();
    expect(props.handleChange).toHaveBeenCalled();

    done();
  });
});
