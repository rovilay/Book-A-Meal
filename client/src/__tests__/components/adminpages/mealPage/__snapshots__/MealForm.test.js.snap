// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MealForm component test should call \`addMeal\` on form submit if not in edit state 1`] = `
<AccordionItemWrapper
  accordionStore={
    AccordionContainer {
      "_listeners": Array [],
      "addItem": [Function],
      "removeItem": [Function],
      "setAccordion": [Function],
      "setExpanded": [Function],
      "setOnChange": [Function],
      "state": Object {
        "accordion": true,
        "items": Array [],
        "onChange": [Function],
      },
    }
  }
  className="accordion__item"
  disabled={false}
  expanded={false}
  hideBodyClassName=""
>
  <AccordionItemTitleWrapper
    className="accordion__title"
    hideBodyClassName=""
  >
    <h3
      className="form-title"
      id="form-title"
    >
      Add Meal
    </h3>
    <FontAwesome
      name="chevron-down"
    />
  </AccordionItemTitleWrapper>
  <AccordionItemBodyWrapper
    className="accordion__body"
    hideBodyClassName="accordion__body--hidden"
  >
    <div
      className="card"
    >
      <div
        className="card-body"
        id="form-card"
      >
        <form
          className="meal-form"
          id="meal-form"
          onSubmit={[Function]}
        >
          <div
            className="name"
          >
            <label
              htmlFor="name"
            >
              Name
            </label>
            <p
              className="input-div"
            >
              <input
                id="meal-name"
                name="title"
                onChange={[Function]}
                placeholder="Enter meal name"
                required={true}
                type="text"
                value="EBA AND EGUSI"
              />
            </p>
          </div>
          <div
            className="price"
          >
            <label
              htmlFor="price"
            >
              Unit Price (₦)
            </label>
            <p
              className="input-div"
            >
              <input
                id="price"
                max="1000000"
                min="1"
                name="price"
                onChange={[Function]}
                placeholder="Enter price"
                required={true}
                type="number"
                value={1000}
              />
            </p>
          </div>
          <div
            className="dsc"
          >
            <label
              htmlFor="dsc"
            >
              Description
            </label>
            <p
              className="input-div desc"
            >
              <input
                id="dsc"
                maxLength="100"
                name="description"
                onChange={[Function]}
                placeholder="Describe meal in not more than 100 words."
                required={true}
                type="text"
                value="So sweet"
              />
            </p>
          </div>
          <div>
            <label
              htmlFor="image"
            >
              Image
            </label>
            <p
              className="input-div"
            >
              <input
                accept="image/*"
                defaultValue=""
                id="meal-image"
                name="image"
                onChange={[MockFunction]}
                placeholder="Enter img link"
                type="file"
              />
            </p>
          </div>
          <div
            className="progress"
          >
            <p
              className="progressBar"
              id="progressBar"
            >
              0%
            </p>
          </div>
          <p
            className="uploadedImage"
          >
            <img
              alt="name"
              src="http://image.com"
            />
          </p>
          <p
            className="buttons"
          >
            <button
              className="btn-2"
              disabled={false}
              id="add-btn"
              name="addbtn"
              type="submit"
            >
              <FontAwesome
                name="plus"
              />
              ADD MEAL
            </button>
          </p>
        </form>
      </div>
    </div>
  </AccordionItemBodyWrapper>
</AccordionItemWrapper>
`;

exports[`MealForm component test should call \`handleChange\` on form input change 1`] = `
<AccordionItemWrapper
  accordionStore={
    AccordionContainer {
      "_listeners": Array [],
      "addItem": [Function],
      "removeItem": [Function],
      "setAccordion": [Function],
      "setExpanded": [Function],
      "setOnChange": [Function],
      "state": Object {
        "accordion": true,
        "items": Array [],
        "onChange": [Function],
      },
    }
  }
  className="accordion__item"
  disabled={false}
  expanded={false}
  hideBodyClassName=""
>
  <AccordionItemTitleWrapper
    className="accordion__title"
    hideBodyClassName=""
  >
    <h3
      className="form-title"
      id="form-title"
    >
      Add Meal
    </h3>
    <FontAwesome
      name="chevron-down"
    />
  </AccordionItemTitleWrapper>
  <AccordionItemBodyWrapper
    className="accordion__body"
    hideBodyClassName="accordion__body--hidden"
  >
    <div
      className="card"
    >
      <div
        className="card-body"
        id="form-card"
      >
        <form
          className="meal-form"
          id="meal-form"
          onSubmit={[Function]}
        >
          <div
            className="name"
          >
            <label
              htmlFor="name"
            >
              Name
            </label>
            <p
              className="input-div"
            >
              <input
                id="meal-name"
                name="title"
                onChange={[Function]}
                placeholder="Enter meal name"
                required={true}
                type="text"
                value="EBA AND EGUSI"
              />
            </p>
          </div>
          <div
            className="price"
          >
            <label
              htmlFor="price"
            >
              Unit Price (₦)
            </label>
            <p
              className="input-div"
            >
              <input
                id="price"
                max="1000000"
                min="1"
                name="price"
                onChange={[Function]}
                placeholder="Enter price"
                required={true}
                type="number"
                value={1000}
              />
            </p>
          </div>
          <div
            className="dsc"
          >
            <label
              htmlFor="dsc"
            >
              Description
            </label>
            <p
              className="input-div desc"
            >
              <input
                id="dsc"
                maxLength="100"
                name="description"
                onChange={[Function]}
                placeholder="Describe meal in not more than 100 words."
                required={true}
                type="text"
                value="So sweet"
              />
            </p>
          </div>
          <div>
            <label
              htmlFor="image"
            >
              Image
            </label>
            <p
              className="input-div"
            >
              <input
                accept="image/*"
                defaultValue=""
                id="meal-image"
                name="image"
                onChange={[MockFunction]}
                placeholder="Enter img link"
                type="file"
              />
            </p>
          </div>
          <div
            className="progress"
          >
            <p
              className="progressBar"
              id="progressBar"
            >
              0%
            </p>
          </div>
          <p
            className="uploadedImage"
          >
            <img
              alt="name"
              src="http://image.com"
            />
          </p>
          <p
            className="buttons"
          >
            <button
              className="btn-2"
              disabled={false}
              id="add-btn"
              name="addbtn"
              type="submit"
            >
              <FontAwesome
                name="plus"
              />
              ADD MEAL
            </button>
          </p>
        </form>
      </div>
    </div>
  </AccordionItemBodyWrapper>
</AccordionItemWrapper>
`;

exports[`MealForm component test should render MealForm correctly 1`] = `
<AccordionItemWrapper
  accordionStore={
    AccordionContainer {
      "_listeners": Array [],
      "addItem": [Function],
      "removeItem": [Function],
      "setAccordion": [Function],
      "setExpanded": [Function],
      "setOnChange": [Function],
      "state": Object {
        "accordion": true,
        "items": Array [],
        "onChange": [Function],
      },
    }
  }
  className="accordion__item"
  disabled={false}
  expanded={false}
  hideBodyClassName=""
>
  <AccordionItemTitleWrapper
    className="accordion__title"
    hideBodyClassName=""
  >
    <h3
      className="form-title"
      id="form-title"
    >
      Edit Meal
    </h3>
    <FontAwesome
      name="chevron-down"
    />
  </AccordionItemTitleWrapper>
  <AccordionItemBodyWrapper
    className="accordion__body"
    hideBodyClassName="accordion__body--hidden"
  >
    <div
      className="card"
    >
      <div
        className="card-body"
        id="form-card"
      >
        <form
          className="meal-form"
          id="meal-form"
          onSubmit={[Function]}
        >
          <div
            className="name"
          >
            <label
              htmlFor="name"
            >
              Name
            </label>
            <p
              className="input-div"
            >
              <input
                id="meal-name"
                name="title"
                onChange={[Function]}
                placeholder="Enter meal name"
                required={true}
                type="text"
                value="EBA AND EGUSI"
              />
            </p>
          </div>
          <div
            className="price"
          >
            <label
              htmlFor="price"
            >
              Unit Price (₦)
            </label>
            <p
              className="input-div"
            >
              <input
                id="price"
                max="1000000"
                min="1"
                name="price"
                onChange={[Function]}
                placeholder="Enter price"
                required={true}
                type="number"
                value={1000}
              />
            </p>
          </div>
          <div
            className="dsc"
          >
            <label
              htmlFor="dsc"
            >
              Description
            </label>
            <p
              className="input-div desc"
            >
              <input
                id="dsc"
                maxLength="100"
                name="description"
                onChange={[Function]}
                placeholder="Describe meal in not more than 100 words."
                required={true}
                type="text"
                value="So sweet"
              />
            </p>
          </div>
          <div>
            <label
              htmlFor="image"
            >
              Image
            </label>
            <p
              className="input-div"
            >
              <input
                accept="image/*"
                defaultValue=""
                id="meal-image"
                name="image"
                onChange={[MockFunction]}
                placeholder="Enter img link"
                type="file"
              />
            </p>
          </div>
          <div
            className="progress"
          >
            <p
              className="progressBar"
              id="progressBar"
            >
              0%
            </p>
          </div>
          <p
            className="uploadedImage"
          >
            <img
              alt="name"
              src="http://image.com"
            />
          </p>
          <p
            className="buttons edit-buttons"
          >
            <button
              className="btn-2"
              disabled={false}
              id="update-btn"
              name="updatebtn"
              type="submit"
            >
              <FontAwesome
                name="arrow-circle-up"
              />
              UPDATE MEAL
            </button>
            <button
              className="btn-2 back-btn"
              disabled={false}
              id="back-btn"
              name="backbtn"
              onClick={[Function]}
              type="submit"
            >
              <FontAwesome
                name="arrow-circle-left"
              />
              BACK
            </button>
          </p>
        </form>
      </div>
    </div>
  </AccordionItemBodyWrapper>
</AccordionItemWrapper>
`;
